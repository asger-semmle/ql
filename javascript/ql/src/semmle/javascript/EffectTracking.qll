/**
 * Provides classes and predicates for reasoning about the ordering of side effects
 * on a value.
 */
import javascript

module EffectTracking {
  private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps

  abstract class Configuration extends string {
    bindingset[this]
    Configuration() { any() }

    /**
     * Holds if the operation at the given location has an effect on the given value,
     * in a way that matters for this effect tracking configuration.
     */
    abstract predicate isTrackedSideEffect(ControlFlowNode loc, DataFlow::SourceNode value);

    /**
     * Holds if a function call itself should be marked as a side effect if the
     * callee has side effects.
     */
    predicate shouldSummarizeCalls() { any() }

    /**
     * Holds if the operation at the given location has an effect on the given value,
     * or (transitively) calls a function that performs side effects on the value.
     */
    predicate hasSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      isTrackedSideEffect(loc, value) or
      calleeHasSideEffects(DataFlow::valueNode(loc), value)
    }

    /**
     * Holds if the given function has a side effect that manipulates `value`.
     */
    predicate functionHasSideEffects(Function f, DataFlow::SourceNode value) {
      exists (ControlFlowNode loc |
        hasSideEffect(loc, value) and
        loc.getContainer() = f)
    }

    /**
     * Holds if the function invoked by `invoke` performs side effects on the given value
     * passed into the call.
     */
    predicate calleeHasSideEffects(DataFlow::InvokeNode invoke, DataFlow::SourceNode value) {
      shouldSummarizeCalls() and
      (
        // Passed as argument
        exists (DataFlow::Node arg, Function target, Parameter param |
          FlowSteps::argumentPassing(invoke, arg, target, param) and
          value.flowsTo(arg) and
          functionHasSideEffects(target, DataFlow::parameterNode(param)))
        or
        // Captured by nested function
        exists (Function target |
          target = invoke.getACallee() and
          functionHasSideEffects(target, value))
      )
    }

    /**
     * Holds if the given basic block has a effect on the given value.
     */
    private predicate basicBlockHasSideEffects(ReachableBasicBlock block, DataFlow::SourceNode value) {
      exists (ControlFlowNode loc | 
        loc.getBasicBlock() = block and
        hasSideEffect(loc, value))
    }

    /**
     * Gets the last side effect to reach `loc`, unless `loc` is an internal node in a basic block
     * without any side effects.
     *
     * In the latter case, the last effect should be obtained using `getSummarizedSideEffectInBasicBlock`.
     */
    private ControlFlowNode getLastSideEffectAtInternalNode(ControlFlowNode loc, DataFlow::SourceNode value) {
      if hasSideEffect(loc, value) then
        result = loc
      else if isStartOfBasicBlock(loc) then
        exists (ControlFlowNode pred | pred = loc.getAPredecessor() |
          if basicBlockHasSideEffects(pred.getBasicBlock(), value) then
            result = getLastSideEffectAtInternalNode(pred, value)
          else
            result = getLastSideEffectAtInternalNode(pred.getBasicBlock().getFirstNode(), value))
      else if basicBlockHasSideEffects(loc.getBasicBlock(), value) then
        result = getLastSideEffectAtInternalNode(loc.getAPredecessor(), value)
      else
        none() // don't store the recent effect on internal nodes
    }

    /**
     * Gets the last side effect to reach the given basic block, assuming the block has no side effects of its own.
     */
    private ControlFlowNode getSummarizedSideEffectInBasicBlock(BasicBlock block, DataFlow::SourceNode value) {
      result = getLastSideEffectAtInternalNode(block.getFirstNode(), value)
      and not basicBlockHasSideEffects(block, value)
    }

    /**
     * Gets the last side effect to affect the given value at the given location.
     */
    pragma[inline]
    ControlFlowNode getLastSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      result = getLastSideEffectAtInternalNode(loc, value)
      or
      result = getSummarizedSideEffectInBasicBlock(loc.getBasicBlock(), value)
    }

    /**
     * Gets the last operation to affect the given value before the given function returns.
     */
    ControlFlowNode getLastSideEffectInFunction(Function f, DataFlow::SourceNode value) {
      result = getLastSideEffect(f.getExit(), value)
    }

    /**
     * Gets the last operation to affect the given value inside the callee of `invoke`.
     *
     * `value` refers to the value passed to the call, and `calleeAlias` refers to its local alias in the callee.
     */
    ControlFlowNode getLastSideEffectInCallee(DataFlow::InvokeNode invoke, DataFlow::SourceNode value, DataFlow::SourceNode calleeAlias) {
      shouldSummarizeCalls() and
      (
        // Passed as argument
        exists (Function target, Parameter parm, DataFlow::Node arg |
          FlowSteps::argumentPassing(invoke, arg, target, parm) and
          value.flowsTo(arg) and
          calleeAlias = DataFlow::parameterNode(parm) and
          result = getLastSideEffectInFunction(target, calleeAlias))
        or
        // Returned from function
        exists (Function target |
          target = invoke.getACallee() and
          value = invoke and
          calleeAlias = target.getAReturnedExpr().flow().getALocalSource() and
          result = getLastSideEffectInFunction(target, calleeAlias))
        or
        // Captured by nested function
        exists (Function target |
          target = invoke.getACallee() and
          value.getContainer() != target and
          result = getLastSideEffectInFunction(target, value) and
          calleeAlias = value)
      )
    }
  }

  private predicate isStartOfBasicBlock(ControlFlowNode node) {
    exists (BasicBlock b | b.getFirstNode() = node)
  }
}
