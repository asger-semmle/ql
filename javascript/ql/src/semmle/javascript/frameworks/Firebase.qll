import javascript

module Firebase {

  private DataFlow::SourceNode getANonLocalSuccessor(DataFlow::SourceNode node) {
    exists (DataFlow::ClassNode class_, string prop |
      class_.getConstructor().getReceiver().getAPropertySource(prop) = node and
      result = class_.getAnInstanceMethod().getReceiver().getAPropertyRead(prop)
    )
  }

  /** Gets a reference to the firebase API object. */
  private DataFlow::SourceNode firebase() {
    result = DataFlow::moduleImport("firebase/app")
    or
    result = DataFlow::globalVarRef("firebase")
    or
    result = getANonLocalSuccessor(firebase())
  }

  /** Gets a reference to a firebase app. */
  DataFlow::SourceNode app() {
    result = firebase()
    or
    result = firebase().getAMethodCall("initializeApp")
    or
    result = getANonLocalSuccessor(app())
  }

  /** Gets a reference to a firebase database object, such as `firebase.database()`. */
  DataFlow::SourceNode database() {
    result = app().getAMethodCall("database")
    or
    result = getANonLocalSuccessor(database())
  }

  /** Gets a call to `ref` or `refFromURL` on a firebase database. */
  DataFlow::SourceNode refSource() {
    exists (string name | result = database().getAMethodCall(name) |
      name = "ref" or
      name = "refFromURL"
    )
  }
  
  private predicate hasFirebaseTaintSource() {
    exists(refSource())
  }

  private class FirebaseRefSource extends DataFlow::AdditionalSource {
    FirebaseRefSource() {
      this = refSource()
    }

    override predicate isSourceFor(DataFlow::Configuration config, DataFlow::FlowLabel label) {
      label = "firebase-ref"
    }
  }
  
  /**
   * Label representing a Firebase reference or a "value" event response from such a reference.
   */
  class FirebaseRefLabel extends DataFlow::FlowLabel {
    FirebaseRefLabel() {
      hasFirebaseTaintSource() and
      this = "firebase-ref"
    }
  }

  /**
   * A step `ref -> ref.val()`, transforming a firebase ref to a tainted value.
   */
  class ValStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
    ValStep() {
      hasFirebaseTaintSource() and
      getMethodName() = "val" and
      getNumArgument() = 0
    }
  
    override predicate step(
      DataFlow::Node pred, DataFlow::Node succ, DataFlow::FlowLabel predlbl,
      DataFlow::FlowLabel succlbl
    ) {
      predlbl = "firebase-ref" and
      succlbl = DataFlow::FlowLabel::taint() and
      pred = getReceiver() and
      succ = this
    }
  }

  /**
   * A step `ref -> ref.child()` or `ref -> ref.push()`, returning a new firebase ref.
   */
  class RefChildStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
    RefChildStep() {
      hasFirebaseTaintSource() and
      exists (string name | name = getMethodName() |
        name = "child" and
        getNumArgument() = 1
        or
        name = "push" and
        getNumArgument() = 0
      )
    }
  
    override predicate step(
      DataFlow::Node pred, DataFlow::Node succ, DataFlow::FlowLabel predlbl,
      DataFlow::FlowLabel succlbl
    ) {
      predlbl = "firebase-ref" and
      succlbl = "firebase-ref" and
      pred = getReceiver() and
      succ = this
    }
  }

  /**
   * A step from `x` to `y` in `x.on("value", y => ...)`.
   */
  class RefListenerStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
    RefListenerStep() {
      hasFirebaseTaintSource() and
      (getMethodName() = "on" or getMethodName() = "once") and
      getArgument(0).asExpr().getStringValue() = "value"
    }
  
    override predicate step(
      DataFlow::Node pred, DataFlow::Node succ, DataFlow::FlowLabel predlbl,
      DataFlow::FlowLabel succlbl
    ) {
      predlbl = "firebase-ref" and
      succlbl = "firebase-ref" and
      pred = getReceiver() and
      (
        succ = this // returns promise
        or
        succ = getCallback(1).getParameter(0)
      )
    }
  }

  /**
   * A step from `x` to `y` in `x.then(y => {})`.
   */
  class RefPromiseStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
    DataFlow::Node dst;

    RefPromiseStep() {
      hasFirebaseTaintSource() and
      promiseTaintStep(this, dst)
    }
  
    override predicate step(
      DataFlow::Node pred, DataFlow::Node succ, DataFlow::FlowLabel predlbl,
      DataFlow::FlowLabel succlbl
    ) {
      predlbl = "firebase-ref" and
      succlbl = "firebase-ref" and
      pred = this and
      succ = dst
    }
  }
}
