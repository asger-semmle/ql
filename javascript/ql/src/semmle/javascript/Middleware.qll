/**
 * Provides generic concepts for modelling the middleware stack of a web application.
 */
import javascript
import semmle.javascript.EffectTracking
import semmle.javascript.frameworks.ExpressMiddleware // imported from here to avoid empty recursion error

module Middleware {
  /**
   * An effect-tracking configuration, tracking installations of route handlers on a router.
   */
  class SetupTracking extends EffectTracking::Configuration {
    SetupTracking() { this = "MiddlewareSetup" }
  
    override predicate isTrackedSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      exists (Setup setup |
        loc = setup.getControlFlowNode() and
        value = setup.getRouter())
    }
  }

  /**
   * A node in a web application's middleware stack model.
   *
   * A node very loosely represents a thing that can handle requests.
   * Incoming requests come from the parent and will be routed to the children,
   * in order of first to last.
   */
  abstract class Node extends DataFlow::Node {
    final ControlFlowNode getControlFlowNode() {
      result = getAstNode()
    }

    /**
     * Gets the local source of the router whose middleware stack is modeled by this node.
     *
     * For routers, this refers to the node itself.
     *
     * For setups that imperatively install route handlers on an existing router, this
     * refers to the router that is being modified.
     */
    abstract DataFlow::SourceNode getRouter();

    /**
     * Gets the last child of this node.
     */
    abstract Node getLastChild();

    /**
     * Gets the previous sibling of this node, that is, the router handler executing
     * before this one.
     *
     * In general there can be multiple previous siblings, forming a sibling graph,
     * but in practice there is rarely more than once.
     */
    abstract Node getPreviousSibling();

    /**
     * Gets the next sibling of this node, that is, the router handler executing
     * after this one.
     *
     * In general there can be multiple next siblings, forming a sibling graph,
     * but in practice there is rarely more than once.
     */
    final Node getNextSibling() {
      this = result.getPreviousSibling()
    }

    /**
     * Gets a child of this node, that is, a router handler that is executed
     * as part of this route handler.
     */
    final Node getAChild() {
      result = this.getLastChild().getPreviousSibling*()
    }
    
    /**
     * Gets the parent of this node, that is, the route handler that dispatches
     * requests to this one.
     */
    final Node getParent() {
      this = result.getAChild()
    }

    /**
     * Gets the path matched by this route handler, relative to its parent.
     *
     * A route handler will reject a request not matching its path.
     *
     * Defaults to the empty string, meaning any request is accepted.
     */
    string getPath() {
      result = ""
    }

    /**
     * Normalizes a path so multiple paths can be joined by concatenation.
     *
     * In particular, this ensures we don't get duplicate `//` path separators.
     */
    bindingset[str]
    private string normalizePath(string str) {
      if str = "/" or str = "" then
        result = ""
      else if str = any(string s) + "/" then
        result = str
      else
        result = str + "/"
    }

    /**
     * Gets the path handled by this node, relative to its parent.
     */
    private string getRelativePath() {
      result = normalizePath(this.getPath())
      or
      not exists(this.getPath()) and
      result = ""
    }

    /**
     * Gets the path handled by this node, relative to the root node.
     */
    final string getAbsolutePath() {
      isTree() and
      (
        result = getParent().getAbsolutePath() + getRelativePath()
        or
        not exists(getParent()) and
        result = getRelativePath()
      )
      or
      // If there are multiple parents, use our best approximation: the absolute paths of any of the parents
      not isTree() and
      result = getParent().getAbsolutePath()
    }

    /**
     * Gets the request method matched by this route handler, or `*` if all
     * request methods are matched.
     *
     * The route handler will reject requests whose request method does not match.
     *
     * Defaults to `*`.
     */
    string getRequestMethod() {
      result = "*"
    }

    /**
     * Gets the request method matched by this router handler and its parents.
     */
    final string getAbsoluteRequestMethod() {
      isRoot() and result = getRequestMethod()
      or
      result = intersectRequestMethodFilters(getParent().getRequestMethod(), getRequestMethod())
    }

    /**
     * Holds if this has no parent.
     */
    final predicate isRoot() {
      not exists(getParent())
    }

    /**
     * Holds if this is part of a tree-shaped middleware stack, which is the common case.
     */
    final predicate isTree() {
      isRoot()
      or
      strictcount(getParent()) = 1 and
      getParent().isTree()
    }

    /**
     * Holds if this handler is preceded by `guard` in the middleware stack.
     *
     * That is, any request seen by this handler is known to have gone through `guard` first.
     */
    final predicate isGuardedBy(Node guard) {
      getParent*().getPreviousSibling+().getAChild*() = guard and
      getAbsolutePath() = guard.getAbsolutePath() + any(string s) and
      exists(intersectRequestMethodFilters(getAbsoluteRequestMethod(), guard.getAbsoluteRequestMethod()))
    }
  }

  /**
   * Gets the intersection of two request method filters.
   */
  bindingset[method1, method2]
  private string intersectRequestMethodFilters(string method1, string method2) {
    method1 = method2 and result = method1
    or
    method1 = "*" and result = method2
    or
    method2 = "*" and result = method1
  }

  /**
   * A middleware node with an indexed list of children.
   *
   * For example, a route setup `app.get("/", handler1, handler2, ...)` has a list of
   * children derived from its argument list.
   */
  private abstract class SequenceNode extends Node {
    /**
     * Gets the `n`th middleware function installed at this point.
     */
    abstract DataFlow::Node getMiddleware(int n);

    /**
     * Gets any of the middleware functions installed at this point.
     */
    final DataFlow::Node getAMiddleware() { result = getMiddleware(_) }

    final int getNumMiddleware() {
      result = count(getAMiddleware())
    }

    final DataFlow::Node getLastMiddleware() {
      result = getMiddleware(getNumMiddleware() - 1)
    }

    override final Node getLastChild() {
      result = getMiddleware(getNumMiddleware() - 1)
    }
  }

  /**
   * An argument in a sequence node.
   */
  private class SequenceElement extends Node {
    SequenceNode parent;
    int index;

    SequenceElement() {
      this = parent.getMiddleware(index)
    }

    override DataFlow::SourceNode getRouter() {
      result = parent.getRouter()
    }
    
    override Node getPreviousSibling() {
      result = parent.getMiddleware(index - 1)
    }
    
    override Node getLastChild() {
      result != this and
      (
        result.(DataFlow::SourceNode).flowsTo(this)
        or
        result.(DataFlow::TrackedNode).flowsTo(this)
      )
    }
  }

  /**
   * An installation of one or more route handlers on a router.
   *
   * The router handler arguments are the children of this node.
   */
  abstract class Setup extends SequenceNode {
    override final Node getPreviousSibling() {
      result.getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(getControlFlowNode().getAPredecessor(), getRouter())
    }
  }

  /**
   * A router, that is, a collection of route handlers with associated paths and request methods.
   */
  class Router extends Node {
    Router() {
      this = any(Node node).getRouter()
    }

    override DataFlow::SourceNode getRouter() {
      result = this
    }
    
    override Node getPreviousSibling() {
      none()
    }
    
    override Node getLastChild() {
      result.(Setup).getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(this.getContainer().getExit(), this)
    }
  }

  /**
   * A functionally defined router, as opposed to routers built up imperatively by adding router handlers on them.
   *
   * For example, an array of route handlers can in some frameworks be used as a route handler itself.
   * In this case, an array literal `[r1, r2]` would be considered a route combinator, with `r1` and `r2` as
   * its children.
   *
   * A route combinator is a `Setup` with themselves as the router.
   */
  abstract class RouteCombinator extends SequenceNode {
    override DataFlow::SourceNode getRouter() {
      result = this
    }

    override Middleware::Node getPreviousSibling() {
      none()
    }
  }

  /**
   * An array literal passed to a route setup.
   */
  private class ArrayRouteCombinator extends Middleware::RouteCombinator, DataFlow::ArrayCreationNode {
    ArrayRouteCombinator() {
      this.flowsTo(any(Node node))
    }
  
    override DataFlow::Node getMiddleware(int n) {
      result = getElement(n)
    }
  }

  /**
   * A call to a function that contains middleware installations.
   */
  class SubSetup extends Middleware::Node, DataFlow::InvokeNode {
    DataFlow::SourceNode router;

    SubSetup() {
      any(SetupTracking tr).calleeHasSideEffects(this, router)
    }

    override DataFlow::SourceNode getRouter() {
      result = router
    }

    override Middleware::Node getLastChild() {
      exists (DataFlow::SourceNode subrouter, ControlFlowNode loc |
        loc = any(SetupTracking tr).getLastSideEffectInCallee(this, router, subrouter) and
        if subrouter = router then
          result.getControlFlowNode() = loc
        else
          result = subrouter)
    }

    override Middleware::Node getPreviousSibling() {
      result.getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(getControlFlowNode().getAPredecessor(), router)
    }
  }
}
