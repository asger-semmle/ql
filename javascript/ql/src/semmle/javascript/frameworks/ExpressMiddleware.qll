import javascript
import semmle.javascript.Middleware

/**
 * An Express route setup as a Middleware setup.
 */
private class ExpressMiddlewareSetup extends Middleware::Setup {
  Express::RouteSetup setup;

  ExpressMiddlewareSetup() {
    this = setup.flow()
  }

  override DataFlow::SourceNode getRouter() {
    result = setup.getReceiver().flow().getALocalSource()
  }

  override DataFlow::Node getMiddleware(int n) {
    result = setup.getRouteHandlerExpr(n).flow()
  }

  override string getPath() {
    result = setup.getPath()
  }
  
  override string getRequestMethod() {
    result = setup.getRequestMethod()
    or
    setup.handlesAllRequestMethods() and
    result = "*"
  }
}
