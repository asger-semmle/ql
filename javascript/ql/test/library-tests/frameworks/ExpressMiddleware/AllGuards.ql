import javascript
import semmle.javascript.frameworks.ExpressMiddleware

from Middleware::Node node, Middleware::Node guard
where node.isGuardedBy(guard)
select node, guard

