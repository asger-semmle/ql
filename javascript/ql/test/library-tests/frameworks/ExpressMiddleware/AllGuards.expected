| tst.js:5:31:5:33 | app | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:5:31:5:33 | app | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:6:3:6:33 | app.get ... are(1)) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:6:3:6:33 | app.get ... are(1)) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:6:16:6:32 | makeMiddleware(1) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:6:16:6:32 | makeMiddleware(1) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:10:3:10:33 | app.get ... are(2)) | tst.js:5:31:5:33 | app |
| tst.js:10:3:10:33 | app.get ... are(2)) | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:10:3:10:33 | app.get ... are(2)) | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:10:3:10:33 | app.get ... are(2)) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:10:3:10:33 | app.get ... are(2)) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:10:3:10:33 | app.get ... are(2)) | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:10:16:10:32 | makeMiddleware(2) | tst.js:5:31:5:33 | app |
| tst.js:10:16:10:32 | makeMiddleware(2) | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:10:16:10:32 | makeMiddleware(2) | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:10:16:10:32 | makeMiddleware(2) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:10:16:10:32 | makeMiddleware(2) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:10:16:10:32 | makeMiddleware(2) | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:14:16:14:31 | express.Router() | tst.js:5:31:5:33 | app |
| tst.js:14:16:14:31 | express.Router() | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:14:16:14:31 | express.Router() | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:14:16:14:31 | express.Router() | tst.js:10:3:10:33 | app.get ... are(2)) |
| tst.js:14:16:14:31 | express.Router() | tst.js:10:16:10:32 | makeMiddleware(2) |
| tst.js:14:16:14:31 | express.Router() | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:14:16:14:31 | express.Router() | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:14:16:14:31 | express.Router() | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:14:16:14:31 | express.Router() | tst.js:22:1:22:21 | setupTh ... pture() |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:5:31:5:33 | app |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:10:3:10:33 | app.get ... are(2)) |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:10:16:10:32 | makeMiddleware(2) |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:15:3:15:36 | router. ... are(3)) | tst.js:22:1:22:21 | setupTh ... pture() |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:5:31:5:33 | app |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:10:3:10:33 | app.get ... are(2)) |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:10:16:10:32 | makeMiddleware(2) |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:15:19:15:35 | makeMiddleware(3) | tst.js:22:1:22:21 | setupTh ... pture() |
| tst.js:21:1:21:25 | setupTh ... nt(app) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:21:1:21:25 | setupTh ... nt(app) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:22:1:22:21 | setupTh ... pture() | tst.js:5:31:5:33 | app |
| tst.js:22:1:22:21 | setupTh ... pture() | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:22:1:22:21 | setupTh ... pture() | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:22:1:22:21 | setupTh ... pture() | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:22:1:22:21 | setupTh ... pture() | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:22:1:22:21 | setupTh ... pture() | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:5:31:5:33 | app |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:10:3:10:33 | app.get ... are(2)) |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:10:16:10:32 | makeMiddleware(2) |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:23:1:23:28 | app.get ... uter()) | tst.js:22:1:22:21 | setupTh ... pture() |
| tst.js:23:14:23:27 | returnRouter() | tst.js:5:31:5:33 | app |
| tst.js:23:14:23:27 | returnRouter() | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:23:14:23:27 | returnRouter() | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:23:14:23:27 | returnRouter() | tst.js:10:3:10:33 | app.get ... are(2)) |
| tst.js:23:14:23:27 | returnRouter() | tst.js:10:16:10:32 | makeMiddleware(2) |
| tst.js:23:14:23:27 | returnRouter() | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:23:14:23:27 | returnRouter() | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:23:14:23:27 | returnRouter() | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:23:14:23:27 | returnRouter() | tst.js:22:1:22:21 | setupTh ... pture() |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:5:31:5:33 | app |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:10:3:10:33 | app.get ... are(2)) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:10:16:10:32 | makeMiddleware(2) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:14:16:14:31 | express.Router() |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:15:3:15:36 | router. ... are(3)) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:15:19:15:35 | makeMiddleware(3) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:22:1:22:21 | setupTh ... pture() |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:23:1:23:28 | app.get ... uter()) |
| tst.js:25:1:25:31 | app.get ... are(4)) | tst.js:23:14:23:27 | returnRouter() |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:5:31:5:33 | app |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:6:3:6:33 | app.get ... are(1)) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:6:16:6:32 | makeMiddleware(1) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:10:3:10:33 | app.get ... are(2)) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:10:16:10:32 | makeMiddleware(2) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:14:16:14:31 | express.Router() |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:15:3:15:36 | router. ... are(3)) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:15:19:15:35 | makeMiddleware(3) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:19:1:19:31 | app.get ... are(0)) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:19:14:19:30 | makeMiddleware(0) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:21:1:21:25 | setupTh ... nt(app) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:22:1:22:21 | setupTh ... pture() |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:23:1:23:28 | app.get ... uter()) |
| tst.js:25:14:25:30 | makeMiddleware(4) | tst.js:23:14:23:27 | returnRouter() |
